datafusion/core/src/physical_optimizer/pipeline_checker.rs

```diff
@@ -0,0 +1,1173 @@
+        )),
+    }
+}
+
+/// [PipelineStatePropagator] propagates the [ExecutionPlan] pipelining information.
+#[derive(Clone, Debug)]
+pub struct PipelineStatePropagator {
+    pub(crate) plan: Arc<dyn ExecutionPlan>,
+    pub(crate) unbounded: bool,
+    pub(crate) children_unbounded: Vec<bool>,
+}
+
+impl PipelineStatePropagator {
+    /// Constructs a new, default pipelining state.
+    pub fn new(plan: Arc<dyn ExecutionPlan>) -> Self {
+        let length = plan.children().len();
+        PipelineStatePropagator {
+            plan,
+            unbounded: false,
+            children_unbounded: vec![false; length],
+        }
+    }
+    /// It generates children of the execution plan with state.
+    pub fn children(&self) -> Vec<PipelineStatePropagator> {
+        self.plan
+            .children()
+            .into_iter()
+            .map(|child| {
+                let length = child.children().len();
+                PipelineStatePropagator {
+                    plan: child,
+                    unbounded: false,
+                    children_unbounded: vec![false; length],
+                }
+            })
+            .collect()
+    }
+}
+
+impl TreeNodeRewritable for PipelineStatePropagator {
+    fn map_children<F>(self, transform: F) -> Result<Self>
+    where
+        F: FnMut(Self) -> Result<Self>,
+    {
+        let children = self.children();
+        if !children.is_empty() {
+            let new_children = children
+                .into_iter()
+                .map(transform)
+                .collect::<Result<Vec<_>>>()?;
+            let children_unbounded = new_children
+                .iter()
+                .map(|c| c.unbounded)
+                .collect::<Vec<bool>>();
+            let children_plans = new_children
+                .into_iter()
+                .map(|child| child.plan)
+                .collect::<Vec<_>>();
+            Ok(PipelineStatePropagator {
+                plan: with_new_children_if_necessary(self.plan, children_plans)?,
+                unbounded: self.unbounded,
+                children_unbounded,
+            })
+        } else {
+            Ok(self)
+        }
+    }
+}
+
+fn apply_subrules_and_check_finiteness_requirements(
+    mut input: PipelineStatePropagator,
+    physical_optimizer_subrules: &Vec<Box<PipelineCheckerSubrule>>,
+) -> Result<Option<PipelineStatePropagator>> {
+    for sub_rule in physical_optimizer_subrules {
+        match sub_rule(&input) {
+            Some(Ok(value)) => {
+                input = value;
+            }
+            Some(Err(e)) => return Err(e),
+            _ => {}
+        }
+    }
+
+    let plan = input.plan;
+    let children = &input.children_unbounded;
+    match plan.unbounded_output(children) {
+        Ok(value) => Ok(Some(PipelineStatePropagator {
+            plan,
+            unbounded: value,
+            children_unbounded: input.children_unbounded,
+        })),
+        Err(e) => Err(e),
+    }
+}
```

Comment:
I think you can express the same logic somewhat more concisely via https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose and https://doc.rust-lang.org/std/result/enum.Result.html#method.map if you wanted

```suggestion
    for sub_rule in physical_optimizer_subrules {
        if let Some(value) = sub_rule(&input).transpose()? {
            input = value;
        }
    }

    let plan = input.plan;
    let children = &input.children_unbounded;
    plan.unbounded_output(children)
        .map(|value| {
            Some(PipelineStatePropagator {
                plan,
                unbounded: value,
                children_unbounded: input.children_unbounded,
            })
        })
}
```